### PR TITLE
feat: add support for configurable registries and applicable authentication options

### PR DESCRIPTION
**Pull request by micsco**
This adds support per #66 

Resolves https://github.com/nodejs/corepack/issues/66

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
Could you do this without using `npm-registry-fetch`?
This PR increases the bundle size by 1MB and that seems a big excessive for essentially adding a few headers.

**Comment by micsco** [comment]
@merceyz - Per your suggestion, I removed the dependencies and manually added the appropriate headers.

Now I see the following results when benchmarking:

```hyperfine.sh-session
$ hyperfine "node ./dist-before/corepack.js --version" "node ./dist-after/corepack.js --version"
Benchmark 1: node ./dist-before/corepack.js --version
  Time (mean ± σ):      55.9 ms ±   2.1 ms    [User: 49.8 ms, System: 7.8 ms]
  Range (min … max):    53.9 ms …  65.2 ms    47 runs

Benchmark 2: node ./dist-after/corepack.js --version
  Time (mean ± σ):      56.2 ms ±   2.1 ms    [User: 49.1 ms, System: 8.8 ms]
  Range (min … max):    53.5 ms …  62.2 ms    54 runs

Summary
  'node ./dist-before/corepack.js --version' ran
    1.00 ± 0.05 times faster than 'node ./dist-after/corepack.js --version'
```

**Comment by xiaoxiangmoe** [comment]
This PR helps me a lot.

File: README.md [hunk 1]
```diff
@@ -129,8 +129,16 @@ This command will retrieve the given package manager from the specified archive
...
+- `COREPACK_NPM_REGISTRY` sets the registry base url used when retrieving package managers from npm. Default value is `https://registry.npmjs.org`
+
+- `COREPACK_NPM_TOKEN` sets a Bearer token authorization header when connecting to `npm` type registry.
+
+- `COREPACK_NPM_USERNAME` and `COREPACK_NPM_PASSWORD` to set a Basic authorization header when connecting to NPM type registry. Note: both environment variables are required, if you want to send an empty password, explicitly set `COREPACK_NPM_PASSWORD` to an empty string.
+
 - `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` are supported through [`node-proxy-agent`](https://github.com/TooTallNate/node-proxy-agent).
 
+[`npm-registry-fetch`]: https://github.com/npm/npm-registry-fetch
+
```
Comment (aduh95) [inline README.md:141]: This is no longer needed, correct?
```suggestion
```

File: sources/npmRegistryUtils.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  if (process.env.COREPACK_ENABLE_NETWORK === `0`)
+    throw new UsageError(`Network access disabled by the environment; can't reach npm repository ${npmRegistryUrl}`);
+
+  const headers = {...DEFAULT_HEADERS};
+
+  if (`COREPACK_NPM_TOKEN` in process.env) {
+    headers.authorization = `Bearer ${process.env.COREPACK_NPM_TOKEN}`;
+  } else if (`COREPACK_NPM_USERNAME` in process.env
+          && `COREPACK_NPM_PASSWORD` in process.env) {
+    const encodedPassword = Buffer.from(`${process.env.COREPACK_NPM_PASSWORD}`, `base64`).toString(`utf8`);
```
Comment (aduh95) [inline sources/npmRegistryUtils.ts:25]: We expect the password to be provided as base64? I think it should be documented somehow.

File: README.md (hunk 1 above)
Comment (micsco) [reply README.md:141]: Correct, updated.

**Comment by micsco** [comment]
Thanks, @aduh95  - Good catch on the password encoding process. 
I've updated both the readme file and the password encoding - I believe that the usual behaviour is just to have the password in plaintext in the environment variables so I have made that so.

File: README.md [hunk 3]
```diff
@@ -129,8 +129,15 @@ This command will retrieve the given package manager from the specified archive
...
 
+- `COREPACK_NPM_REGISTRY` sets the registry base url used when retrieving package managers from npm. Default value is `https://registry.npmjs.org`
+
+- `COREPACK_NPM_TOKEN` sets a Bearer token authorization header when connecting to `npm` type registry.
+
+- `COREPACK_NPM_USERNAME` and `COREPACK_NPM_PASSWORD` to set a Basic authorization header when connecting to NPM type registry. Note: both environment variables are required, as plain text. If you want to send an empty password, explicitly set `COREPACK_NPM_PASSWORD` to an empty string.
+
 - `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` are supported through [`node-proxy-agent`](https://github.com/TooTallNate/node-proxy-agent).
 
+
```
Comment (aduh95) [inline README.md:140]:
```suggestion
```

File: sources/corepackUtils.ts [hunk 4]
```diff
@@ -10,14 +10,14 @@ import * as folderUtils                                        from './folderUti
 import * as fsUtils                                            from './fsUtils';
 import * as httpUtils                                          from './httpUtils';
 import * as nodeUtils                                          from './nodeUtils';
+import * as npmRegistryUtils                                   from './npmRegistryUtils';
 import {RegistrySpec, Descriptor, Locator, PackageManagerSpec} from './types';
 
+
```
Comment (aduh95) [inline sources/corepackUtils.ts:16]:
```suggestion
```

**Review by aduh95 — APPROVED** [review]
LGTM modulo a few unrelated empty lines that were added probably by accident.

File: sources/npmRegistryUtils.ts (hunk 2 above)
Comment (micsco) [reply sources/npmRegistryUtils.ts]: @aduh95 ah good catch, I didn't even think this through - I adopted the same process that was present in the npm flow. I don't actually see the need for this so I've updated it in my latest push.

**Comment by aduh95** [comment]
@micsco it looks like the added test fails:

```
FAIL tests/npmRegistryUtils.test.ts
  ● npm registry utils fetchAsJson › adds authorization header with basic auth if COREPACK_NPM_USERNAME and COREPACK_NPM_PASSWORD are set
    expect(jest.fn()).lastCalledWith(...expected)
    - Expected
    + Received
      "https://registry.npmjs.org/package-name",
      Object {
        "headers": Object {
          "Accept": "application/vnd.npm.install-v1+json",
    -     "authorization": "Basic am9objplYXN5cGFzczEyMw==",
    +     "authorization": "Basic am9objpqb2hu",
        },
      },
    Number of calls: 1
      73 |
      74 |     expect(httpFetchAsJson).toBeCalled();
    > 75 |     expect(httpFetchAsJson).lastCalledWith(`${DEFAULT_NPM_REGISTRY_URL}/package-name`, {headers: {
         |                             ^
      76 |       ...DEFAULT_HEADERS,
      77 |       authorization: `Basic ${encodedCreds}`,
      78 |     }});
      at Object.lastCalledWith (tests/npmRegistryUtils.test.ts:75:29)
```

Can you take a look?

**Comment by micsco** [comment]
@aduh95 apologies, I used the Username env var twice, instead of using the password - thankfully tests caught it but I forgot to check. Have updated the branch.

File: tests/npmRegistryUtils.test.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+  afterEach(() => {
+    process.env = OLD_ENV; // Restore old environment
+  });
+
+  it(`throw usage error if COREPACK_ENABLE_NETWORK env is set to 0`, async () => {
+    process.env.COREPACK_ENABLE_NETWORK = `0`;
+
+    expect(fetchAsJson(`package-name`)).rejects.toThrowError();
```
Comment (merceyz) [inline tests/npmRegistryUtils.test.ts:21]:
```suggestion
    await expect(fetchAsJson(`package-name`)).rejects.toThrowError();
```

**Review by merceyz — CHANGES_REQUESTED** [review]

**Comment by merceyz** [comment]
Could you update the description to mention that this PR `Resolves https://github.com/nodejs/corepack/issues/66`?

File: README.md [hunk 6]
```diff
@@ -129,6 +129,12 @@ This command will retrieve the given package manager from the specified archive
 
 - `COREPACK_ROOT` has no functional impact on Corepack itself; it's automatically being set in your environment by Corepack when it shells out to the underlying package managers, so that they can feature-detect its presence (useful for commands like `yarn init`).
 
+- `COREPACK_NPM_REGISTRY` sets the registry base url used when retrieving package managers from npm. Default value is `https://registry.npmjs.org`
+
+- `COREPACK_NPM_TOKEN` sets a Bearer token authorization header when connecting to `npm` type registry.
+
+- `COREPACK_NPM_USERNAME` and `COREPACK_NPM_PASSWORD` to set a Basic authorization header when connecting to NPM type registry. Note: both environment variables are required, as plain text. If you want to send an empty password, explicitly set `COREPACK_NPM_PASSWORD` to an empty string.
```
Comment (merceyz) [inline README.md:136]:
```suggestion
- `COREPACK_NPM_USERNAME` and `COREPACK_NPM_PASSWORD` to set a Basic authorization header when connecting to npm type registry. Note: both environment variables are required, as plain text. If you want to send an empty password, explicitly set `COREPACK_NPM_PASSWORD` to an empty string.
```
https://github.com/npm/cli/tree/37bf6d31bd2f68e661928744833b57dcabbb0d1a#is-it-npm-or-npm-or-npm

File: sources/npmRegistryUtils.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import {UsageError}          from 'clipanion';
+import {OutgoingHttpHeaders} from 'http2';
+
+import * as httpUtils        from './httpUtils';
+
+// load abbreviated metadata as that's all we need for these calls
+// see: https://github.com/npm/registry/blob/master/docs/responses/package-metadata.md
```
Comment (merceyz) [inline sources/npmRegistryUtils.ts:7]:
```suggestion
// see: https://github.com/npm/registry/blob/cfe04736f34db9274a780184d1cdb2fb3e4ead2a/docs/responses/package-metadata.md
```

File: tests/npmRegistryUtils.test.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    expect(httpFetchAsJson).lastCalledWith(`${DEFAULT_NPM_REGISTRY_URL}/package-name`, {headers: {
+      ...DEFAULT_HEADERS,
+      authorization: `Bearer ${process.env.COREPACK_NPM_TOKEN}`,
+    }});
+  });
+
+  it(`only adds authorization header with bearer token if COREPACK_NPM_TOKEN and COREPACK_NPM_USERNAME are set`, async () => {
+    process.env.COREPACK_NPM_TOKEN = `kcaperoc`;
+    process.env.COREPACK_NPM_USERNAME = `john`;
+    process.env.COREPACK_NPM_PASSWORD = `easypass123`;
```
Comment (merceyz) [inline tests/npmRegistryUtils.test.ts:54]:
```suggestion
    process.env.COREPACK_NPM_TOKEN = `foo`;
    process.env.COREPACK_NPM_USERNAME = `bar`;
    process.env.COREPACK_NPM_PASSWORD = `foobar`;
```

File: tests/npmRegistryUtils.test.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    expect(httpFetchAsJson).lastCalledWith(`${DEFAULT_NPM_REGISTRY_URL}/package-name`, {headers: {
+      ...DEFAULT_HEADERS,
+      authorization: `Bearer ${process.env.COREPACK_NPM_TOKEN}`,
+    }});
+  });
+
+
+  it(`adds authorization header with basic auth if COREPACK_NPM_USERNAME and COREPACK_NPM_PASSWORD are set`, async () => {
+    process.env.COREPACK_NPM_USERNAME = `john`;
+    process.env.COREPACK_NPM_PASSWORD = `easypass123`;
```
Comment (merceyz) [inline tests/npmRegistryUtils.test.ts:68]:
```suggestion
    process.env.COREPACK_NPM_USERNAME = `foo`;
    process.env.COREPACK_NPM_PASSWORD = `bar`;
```

File: tests/npmRegistryUtils.test.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+    expect(httpFetchAsJson).toBeCalled();
+    expect(httpFetchAsJson).lastCalledWith(`${DEFAULT_NPM_REGISTRY_URL}/package-name`, {headers: {
+      ...DEFAULT_HEADERS,
+      authorization: `Basic ${encodedCreds}`,
+    }});
+  });
+
+  it(`does not add authorization header if COREPACK_NPM_USERNAME is set and COREPACK_NPM_PASSWORD is not.`, async () => {
+    process.env.COREPACK_NPM_USERNAME = `john`;
```
Comment (merceyz) [inline tests/npmRegistryUtils.test.ts:82]:
```suggestion
    process.env.COREPACK_NPM_USERNAME = `foo`;
```

File: tests/npmRegistryUtils.test.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  it(`loads from custom COREPACK_NPM_REGISTRY if set`, async () => {
+    process.env.COREPACK_NPM_REGISTRY = `https://registry.example.org`;
+    await fetchAsJson(`package-name`);
+
+    expect(httpFetchAsJson).toBeCalled();
+    expect(httpFetchAsJson).lastCalledWith(`${process.env.COREPACK_NPM_REGISTRY}/package-name`, {headers: DEFAULT_HEADERS});
+  });
+
+  it(`adds authorization header with bearer token if COREPACK_NPM_TOKEN is set`, async () => {
+    process.env.COREPACK_NPM_TOKEN = `kcaperoc`;
```
Comment (merceyz) [inline tests/npmRegistryUtils.test.ts:40]:
```suggestion
    process.env.COREPACK_NPM_TOKEN = `foo`;
```

File: README.md [hunk 12]
```diff
@@ -129,6 +129,12 @@ This command will retrieve the given package manager from the specified archive
 
 - `COREPACK_ROOT` has no functional impact on Corepack itself; it's automatically being set in your environment by Corepack when it shells out to the underlying package managers, so that they can feature-detect its presence (useful for commands like `yarn init`).
 
+- `COREPACK_NPM_REGISTRY` sets the registry base url used when retrieving package managers from npm. Default value is `https://registry.npmjs.org`
+
+- `COREPACK_NPM_TOKEN` sets a Bearer token authorization header when connecting to `npm` type registry.
```
Comment (merceyz) [inline README.md:134]:
```suggestion
- `COREPACK_NPM_TOKEN` sets a Bearer token authorization header when connecting to a `npm` type registry.
```

**Review by merceyz — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (micsco) [reply README.md]: Today I learned! Thank you for the insight as to npm capitalisation

**Comment by micsco** [comment]
Thanks for the comprehensive review @merceyz - I've applied all your changes and updated the PR.

**Pull request merged**